geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/MessageDispatcher.java

```diff
@@ -432,34 +442,37 @@ protected void runDispatcher() {
           _messageQueue.remove();
           clientMessage = null;
         } catch (AuthenticationExpiredException expired) {
-          if (waitForReAuthenticationStartTime == -1) {
-            waitForReAuthenticationStartTime = System.currentTimeMillis();
-            // only send the message to clients who can handle the message
-            if (getProxy().getVersion().isNewerThanOrEqualTo(RE_AUTHENTICATION_START_VERSION)) {
-              EventID eventId = createEventId();
-              sendMessageDirectly(new ClientReAuthenticateMessage(eventId));
+          waitForReAuthenticationStartTime = System.currentTimeMillis();
+          long waitFinishTime = waitForReAuthenticationStartTime + reAuthenticateWaitTime;
+          // only send the message to clients who can handle the message
+          if (getProxy().getVersion().isNewerThanOrEqualTo(RE_AUTHENTICATION_START_VERSION)) {
+            EventID eventId = createEventId();
+            sendMessageDirectly(new ClientReAuthenticateMessage(eventId));
```

Comment:
sendMessageDirectly may be time consuming...Is it a good idea to calculate waitFinishTime after sending the message.

Reply:
done